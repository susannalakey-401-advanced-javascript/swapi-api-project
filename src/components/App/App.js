import React from 'react';
import './App.scss';
import Header from '../Header/Header'
import Results from '../Results/Results'
import Footer from '../Footer/Footer'
import { If, When, Unless } from '../Conditionals/Conditionals'
import loading from './loading.gif'


class App extends React.Component {
  constructor() {
    super()
    this.state = {
      url: 'https://swapi.co/api/people/',
      body: undefined,
      headers: {},
      loading: false,
    };

  }

  getResults = async (method) => {
    this.setState({
      body: undefined,
      headers: {},
    })
    console.log('start')
    this.setState({ loading: true })
    const response = await fetch(this.state.url, { method, mode: 'cors' })
    const body = await response.json()
    this.setState({
      body: body,
      headers: [...response.headers.entries()],
      loading: false,
    })
    this.setLocalStorage();
    console.log('end')
  }

  handleChange = e => {
    this.setState({ url: e.target.value });
  }



  // only runs first time app.js renders
  componentDidMount() {
    const url = new URLSearchParams(window.location.search).get('url');
    if (url) {
      this.setState({
        url
        // put this.getResults in a callback function because it guarantees that the state has been set when we call it
      }, () => {
        this.getResults('GET')
      })


    }
  }


  setLocalStorage = e => {
    // localStorage.setItem('urls', this.state.url)
    const urlsRaw = localStorage.getItem('urls');
    let urls;
    if (urlsRaw) {
      urls = JSON.parse(urlsRaw);
    } else {
      urls = [];
    }
    urls.push(this.state.url);
    localStorage.setItem('urls', JSON.stringify(urls))
  }

  render() {
    return (
      <>
        <div className="App">
          <Header />
          <If condition={this.state.body === null}>
            <p className="fetch-error">Please enter a valid search URL</p>
          </If>

          <form>
            <label>
              Enter API URL:
              <input className="text-input" type="text" value={this.state.url} onChange={this.handleChange} />
            </label>
            <div>

              <button type="button" onClick={() => { this.getResults('GET') }} className="submit">GET</button>
              <button type="button" onClick={() => { this.getResults('PUT') }} className="submit">PUT</button>
              <button type="button" onClick={() => { this.getResults('POST') }} className="submit">POST</button>
              <button type="button" onClick={() => { this.getResults('DELETE') }} className="submit">DELETE</button>
            </div>

          </form>
          <If condition={this.state.loading}>
            <img src={loading} className="loading" />
          </If>
          <If condition={!this.state.isHidden}>
            <Results headers={this.state.headers} body={this.state.body} />
          </If>
          <Footer />
        </div>
      </>
    );
  }
}

export default App;
