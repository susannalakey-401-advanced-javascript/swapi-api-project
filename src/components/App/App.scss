@import url('https://fonts.googleapis.com/css?family=Staatliches&display=swap');

body {
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: rgb(219, 241, 241);
  font-family: 'Staatliches', cursive;
  width: 100%;
}

.App {

  flex-direction: column
}
form {
  .text-input {
    width: 250px;
    height: 20px;
    margin: 40px 20px 20px 20px;
    font-size: 15px;
  }

  button {
    width: 100px;
    height: 30px;
    font-size: 20px;
    margin: 10px;
    height: 40px;
    font-family: 'Staatliches', cursive;
    background: rgb(255, 253, 253);
  }

  label {
    font-size: 22px;
    margin-left: 50px;
  }

}

.fetch-error {
  text-align: center;
  color: red;
}

.loading {
 margin-left: 225px;
 margin-top: 50px;
}