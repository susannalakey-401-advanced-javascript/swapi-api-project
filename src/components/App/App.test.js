import React from 'react'
import App from './App'
import { shallow } from 'enzyme'

describe('<App />', () => {
  let tree;

  beforeEach(() => {
    tree = shallow(<App />)
  })

  it('renders to the page', () => {
    expect(tree).toMatchSnapshot();
  })
})