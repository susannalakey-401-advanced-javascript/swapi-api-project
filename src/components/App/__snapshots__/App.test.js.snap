// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders to the page 1`] = `
<Fragment>
  <div
    className="App"
  >
    <Header />
    <If
      condition={false}
    >
      <p
        className="fetch-error"
      >
        Please enter a valid search URL
      </p>
    </If>
    <form>
      <label>
        Enter API URL:
        <input
          className="text-input"
          onChange={[Function]}
          type="text"
          value="https://swapi.co/api/people/"
        />
      </label>
      <div>
        <button
          className="submit get"
          onClick={[Function]}
          type="button"
        >
          GET
        </button>
        <button
          className="submit"
          onClick={[Function]}
          type="button"
        >
          PUT
        </button>
        <button
          className="submit"
          onClick={[Function]}
          type="button"
        >
          POST
        </button>
        <button
          className="submit"
          onClick={[Function]}
          type="button"
        >
          DELETE
        </button>
      </div>
    </form>
    <If
      condition={false}
    >
      <img
        className="loading"
        src="loading.gif"
      />
    </If>
    <If
      condition={true}
    >
      <Results
        headers={Object {}}
      />
    </If>
    <Footer />
  </div>
</Fragment>
`;
